### PR TITLE
Add support for ZeRO-2/3 and ZeRO-offload in fairscale

### PR DESCRIPTION
**Pull request by sgugger**
# What does this PR do?

This PR adds support for the new `FullyShardedDataParallel` introduced in fairscale. See [this PR](https://github.com/facebookresearch/fairscale/pull/413) for more details.

The PR changes a tiny bit the behavior of the `--sharded_ddp` flag/training argument to support a list of options. You can still use the TrainingArguments class with `sharded_dpp=True` but if launching a script, `--sharded_ddp` has to be replaced with `--sharded_ddp simple`. The `--sharded_ddp` was marked as an experimental API so I think this breaking change is fine if properly documented.

Other values supported are: `zero_dp_2`, `zero_dp_2 offload`, `zero_dp_3` and `zero_dp_3 offload`. To fully take advantage of the `zero_dp_3`/`zero_dp_3 offload` the model passed to the `Trainer` will need to have its internal layers wrapped inside the `FullyShardedDataParallel`, but this out of scope for this particular PR.

For all those new modes, the model simply needs to be wrapped inside `FullyShardedDataParallel` but the optimizer needs to be created after the model wrapping (to get the parameters shards).

Note that:
- `predict_with_generate` does not work with this integration
- `cpu_offload` does not work for now due to the bug mentioned in [this issue](https://github.com/facebookresearch/fairscale/issues/421). Once the issue is fixed, the option should work with the existing code.

One thing to think further is that this integration breaks the usual convention that `self.model` is the original model (`FullyShardedDataParallel` consumes the model to use less memory).

### CODE REVIEW

File: src/transformers/trainer.py [hunk 1]
```diff
@@ -132,10 +133,16 @@
     import torch_xla.distributed.parallel_loader as pl
 
 if is_fairscale_available():
+    import fairscale
     from fairscale.nn.data_parallel import ShardedDataParallel as ShardedDDP
     from fairscale.optim import OSS
     from fairscale.optim.grad_scaler import ShardedGradScaler
 
+    if version.parse(fairscale.__version__) >= version.parse("0.3"):
+        from fairscale.nn.data_parallel import FullyShardedDataParallel as FullyShardedDDP
```
Comment (stas00) [inline src/transformers/trainer.py:142]: I think this may introduce a confusion here, should we stick to DP and not DDP to match the real name? i.e. FullyShardedDP and ShardedDP?

Perhaps change the original flag to reflect that as well? `--sharded_dp`?

Comment (stas00) [reply src/transformers/trainer.py:142]: OK, made a request to make those DDP here:
https://github.com/facebookresearch/fairscale/pull/413#issuecomment-784429352

File: src/transformers/training_args.py [hunk 2]
```diff
@@ -236,9 +236,24 @@ class TrainingArguments:
             When resuming training, whether or not to skip the epochs and batches to get the data loading at the same
             stage as in the previous training. If set to :obj:`True`, the training will begin faster (as that skipping
             step can take a long time) but will not yield the same results as the interrupted training would have.
-        sharded_ddp (:obj:`bool`, `optional`, defaults to :obj:`False`):
+        sharded_ddp (:obj:`bool`, :obj:`str` or :class:`~transformers.trainer_utils.ShardedDDPType`, `optional`, defaults to :obj:`False`):
             Use Sharded DDP training from `FairScale <https://github.com/facebookresearch/fairscale>`__ (in distributed
             training only). This is an experimental feature.
+
+            Can take up to six values:
```
Comment (stas00) [inline src/transformers/training_args.py:243]:
```suggestion
            Can be one of the following values:
```

- clarifying that it's one of them
- the total count is of no useful value to the user

File: src/transformers/training_args.py [hunk 3]
```diff
@@ -236,9 +236,24 @@ class TrainingArguments:
...
+
+            - :obj:`"no"`: for no sharded DataParallelism (default behavior)
+            - :obj:`"simple"`: to use first instance of sharded DDP released by fairscale (:obj:`ShardedDDP`) similar
+              to ZeRO-2.
+            - :obj:`"zero_2"`: to use the second instance of sharded DPP released by fairscale (:obj:`FullyShardedDDP`)
```
Comment (stas00) [inline src/transformers/training_args.py:248]: we are smashing concepts a bit here. ZeRO is a big territory with many features. the 3 stages belong to ZeRO-DP part of ZeRO, so ideally this should be `zero_dp_(1|2|3)` or `zero_dp(1|2|3)`.

This is just a suggestion though, if you strongly feel having just the number is clear enough, that's OK too.

Comment (stas00) [reply src/transformers/training_args.py:248]: Oh, and that's why they call it DP and not DDP, but it's ZeRO-DP.

**Comment by stas00** [comment]
> Other values supported are: `zero2`, `zero2_offload`, `zero3` and `zero3_offload`. To fully take advantage of the `zero3`/`zero3_offload` the model passed to the `Trainer` will need to have its internal layers wrapped inside the `FullyShardedDataParallel`, but this out of scope for this particular PR.

Do you feel it's better to hardcode these combinations and not have a more flexible approach of:
```
--sharded_ddp "zero2;offload;future_option"
```
which would enable adding new features in the future, without needing to create all possible combinations of options which would double every time a new option will be added.

> One thing to think further is that this integration breaks the usual convention that `self.model` is the original model (`FullyShardedDataParallel` consumes the model to use less memory).

Yes, we will need to rethink this - the trainer is getting more a more complex.

**Comment by sgugger** [comment]
> Do you feel it's better to hardcode these combinations and not have a more flexible approach of:
>
> --sharded_ddp "zero2;offload;future_option"

Happy to explore that design as it seems more flexible and less prone to future breaking changes. Will adapt the PR accordingly once we get the wrapper to work.

**Comment by stas00** [comment]
Also we need to make sure that we distinguish between `FullyShardedDataParallel` and `ShardedDataParallel` since as the commentary was made, they aren't quite the same. 

Deepspeed has a `stage` param which goes from 0 to 3. where stage=0 doesn't enable ZeRO, and then each number matches the stage.

For the user's sake perhaps we could make things as similar as possible so it'd be more intuitive for them to switch between fairscale (and eventually pytorch) and deepspeed.

Also note that DeepSpeed exposes other params like the size of buckets, which actually are very important and need to be user-configurable. I won't be surprised that FSDP will also have those configurable down the road - i.e. more params.

File: src/transformers/trainer.py [hunk 4]
```diff
@@ -748,18 +757,18 @@ def _wrap_model(self, model, training=True):
...
-            if self.args.sharded_ddp == ShardedDDPType.SIMPLE:
+            if self.sharded_ddp == ShardedDDPOption.SIMPLE:
                 model = ShardedDDP(model, self.optimizer)
             else:
                 mixed_precision = self.args.fp16
-                cpu_offload = self.args.sharded_ddp in [ShardedDDPType.ZERO_2_OFFLOAD, ShardedDDPType.ZERO_3_OFFLOAD]
-                zero_3 = self.args.sharded_ddp in [ShardedDDPType.ZERO_3, ShardedDDPType.ZERO_3_OFFLOAD]
+                cpu_offload = ShardedDDPOption.OFFLOAD in self.args.sharded_ddp
+                zero_3 = self.sharded_ddp == ShardedDDPOption.ZERO_DP_3
                 # Breaking the self.model convention but I see no way around it for now.
```
Comment (stas00) [inline src/transformers/trainer.py:768]:
```suggestion
                # XXX: Breaking the self.model convention but I see no way around it for now.
```

Perhaps let's add a flag to get back to it later? Not attached it - this is usually how I know to come back to something later.

File: src/transformers/training_args.py [hunk 5]
```diff
@@ -458,8 +456,8 @@ class TrainingArguments:
             "help": "When resuming training, whether or not to skip the first epochs and batches to get to the same training data."
         },
     )
-    sharded_ddp: ShardedDDPType = field(
-        default="no",
+    sharded_ddp: str = field(
+        default="",
```
Comment (stas00) [inline src/transformers/training_args.py:460]: perhaps list the choices here? and perhaps a very small example of combining 2 of them in the value, since it's not a usual pattern - a user might struggle here.

**Comment by sgugger** [comment]
Reworked the API to take your suggestion of list of options into account @stas00. I don't think we have to worry about uniformizing with deepspeed or cleaning more at this stage as:
- this API will evolve in the future (ShardedDataParallel might very well disappear if FullyShardedDataParallel is better, and this might change again on the road to be merged in PyTorch)
- we don't know yet all the options we will have between deepspeed/fairscale/PyTorch
- this is an experimental API and while we won't break it just for fun, we can make slight changes down the road.

File: src/transformers/training_args.py [hunk 6]
```diff
@@ -551,8 +549,18 @@ def __post_init__(self):
...
-            self.sharded_ddp = "simple" if self.sharded_ddp else "no"
-        self.sharded_ddp = ShardedDDPType(self.sharded_ddp)
+            self.sharded_ddp = "simple" if self.sharded_ddp else ""
+        if isinstance(self.sharded_ddp, str):
+            self.sharded_ddp = [ShardedDDPOption(s) for s in self.sharded_ddp.split()]
+        if self.sharded_ddp == [ShardedDDPOption.OFFLOAD]:
+            raise ValueError(
+                "`--sharded_ddp offload` can't work on its own. It needs to be added to `--sharded_ddp zero_dp_2` or "
+                "`--sharded_ddp zero_dp_3`"
+            )
```
Comment (stas00) [inline src/transformers/training_args.py:559]:
```suggestion
        if self.sharded_ddp == [ShardedDDPOption.OFFLOAD]:
            raise ValueError(
                "`--sharded_ddp offload` can't work on its own. It needs to be added to `--sharded_ddp zero_dp_2` or "
                "`--sharded_ddp zero_dp_3`". For example, `--sharded_ddp "zero_dp_2 offload"`.
            )
```

Let's help the user here a bit.

**Review by stas00 — APPROVED** [review]
Awesome work, @sgugger !!!

File: docs/source/main_classes/trainer.rst [hunk 7]
```diff
@@ -268,17 +271,55 @@ For example here is how you could use it for ``run_seq2seq.py`` with 2 GPUs:
...
+    --output_dir output_dir --overwrite_output_dir \
+    --do_train --max_train_samples 500 --num_train_epochs 1 \
+    --dataset_name wmt16 --dataset_config "ro-en" \
+    --task translation_en_to_ro --source_prefix "translate English to Romanian: " \
+    --fp16 --sharded_ddp zero_dp_2
+
+:obj:`zero_dp_2` is an optimized version of the simple wrapper, while :obj:`zero_dp_3` fully shards model weights,
+gradients and optimizer states.
+
+Both are compatible with adding :obj:`cpu_offload` to enable ZeRO-offload (activat it like this: :obj:`--sharded_ddp
```
Comment (stas00) [inline docs/source/main_classes/trainer.rst:304]:
```suggestion
Both are compatible with adding :obj:`cpu_offload` to enable ZeRO-offload (activate it like this: :obj:`--sharded_ddp
```

File: docs/source/main_classes/trainer.rst [hunk 8]
```diff
@@ -268,17 +271,55 @@ For example here is how you could use it for ``run_seq2seq.py`` with 2 GPUs:
...
+"zero_dp_2 cpu_offload"`).
+
+Notes:
+
+- This feature requires distributed training (so multiple GPUs).
+- It is not implemented for TPUs.
+- It works with ``--fp16`` too, to make things even faster.
+- The ``cpu_offload`` additional option require ``--fp16``.
```
Comment (stas00) [inline docs/source/main_classes/trainer.rst:312]:
```suggestion
- The ``cpu_offload`` additional option requires ``--fp16``.
```

File: docs/source/main_classes/trainer.rst [hunk 9]
```diff
@@ -268,17 +271,55 @@ For example here is how you could use it for ``run_seq2seq.py`` with 2 GPUs:
...
+- This is an area of active development, so make sure you have a source install of fairscale to use this feature as
+  some bugs you encounter may have been fixed there already.
+
+Known caveats:
+
+- This feature is incompatible with :obj:`--predict_with_generate` in the run_seq2seq script.
```
Comment (stas00) [inline docs/source/main_classes/trainer.rst:318]:
```suggestion
- This feature is incompatible with :obj:`--predict_with_generate` in the `run_seq2seq.py` script.
```

File: docs/source/main_classes/trainer.rst [hunk 10]
```diff
@@ -268,17 +271,55 @@ For example here is how you could use it for ``run_seq2seq.py`` with 2 GPUs:
...
+- Using :obj:`--sharded_ddp zero_dp_3` requires wrapping each layer of the model in the special container
+  :obj:`FullyShardedDataParallelism` of fairscale. This is not done automatically by any of the example script of the
```
Comment (stas00) [inline docs/source/main_classes/trainer.rst:320]:
```suggestion
  :obj:`FullyShardedDataParallelism` of fairscale. This is not done automatically by any of the example scripts of the
```

File: docs/source/main_classes/trainer.rst [hunk 11]
```diff
@@ -268,17 +271,55 @@ For example here is how you could use it for ``run_seq2seq.py`` with 2 GPUs:
...
+   to the command line arguments, and make sure you have added the distributed launcher ``-m torch.distributed.launch
+   --nproc_per_node=NUMBER_OF_GPUS_YOU_HAVE`` if you haven't been using it already.
+
+For example here is how you could use it for ``run_seq2seq.py`` with 2 GPUs:
+
+.. code-block:: bash
+
+    python -m torch.distributed.launch --nproc_per_node=2 examples/seq2seq/run_seq2seq.py \
+    --model_name_or_path t5-small --per_device_train_batch_size 1   \
+    --output_dir output_dir --overwrite_output_dir \
```
Comment (LysandreJik) [inline docs/source/main_classes/trainer.rst:295]: For some reason this doesn't work in my env because `--overwrite_output_dir` requires a value, but it probably has nothing to do with that PR. Will look into it.

File: src/transformers/training_args.py (hunk 5 above)
Comment (LysandreJik) [reply src/transformers/training_args.py:460]: I agree with @stas00!

**Review by LysandreJik — APPROVED** [review]
Fantastic! Left only nitpicks.

File: docs/source/main_classes/trainer.rst (hunk 11 above)
Comment (LysandreJik) [reply docs/source/main_classes/trainer.rst:295]: I found out where it comes from, it's because I'm using the fish shell. The fish shell parses this differently than bash, and understands that we're passing a space to `--overwrite_output_dir`, hence the error:

```
run_seq2seq.py: error: argument --overwrite_output_dir: Truthy value expected: got  but expected one of yes/no, true/false, t/f, y/n, 1/0 (case insensitive).
```

More of an issue with argparse than with the script, so no problem there.

Comment (LysandreJik) [reply docs/source/main_classes/trainer.rst:295]: Actually I can't even reproduce anymore, nevermind. Resolving.

File: src/transformers/trainer.py (hunk 1 above)
Comment (min-xu-ai) [reply src/transformers/trainer.py:142]: Thanks Stas. I personally think the distinction between DDP and DP is not going to matter anymore. Even pytorch DDP itself is moving to remove the "device_ids" argument in the future so that there isn't a support for a single process DP (as opposed to distributed/multiprocess DP). Therefore, I think sticking with FSDP is fine within fairscale.

Comment (stas00) [reply src/transformers/trainer.py:142]: Thank you for your follow up, @min-xu-ai

**Comment by stas00** [comment]
Moving out the cl arg naming discussion from https://github.com/huggingface/transformers/pull/10354#pullrequestreview-596676591 to the open

Then we should probably change the cl arg to `_dp` as I suggested above so that it's consistently either DP or DDP all the way through.

@Sylvain, perhaps we should just call it `--sharded`? the the dp part is already inside the value anyway as in: `--sharded zero_dp_3`
